### PR TITLE
Migrate Deprecated Transform Methods to Process Methods

### PR DESCRIPTION
**Pull request by hrishabhg**
### Description 
KStream has deprecated `transformer` methods in earlier versions and have removed them in ce-kafka:8.0.

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")
- [ ] Do these changes have compatibility implications for rollback? If so, ensure that the ksql [command version](https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/Command.java#L41) is bumped.

### CODE REVIEW

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/RestIntegrationTestUtil.java [hunk 1]
```diff
@@ -65,6 +65,7 @@
 import io.vertx.ext.web.client.WebClientOptions;
 import io.vertx.ext.web.codec.BodyCodec;
 import java.net.URI;
+import java.net.URLEncoder;
```
Comment (mjsax) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/RestIntegrationTestUtil.java:68]: Why is this new import required?

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SinkBuilder.java [hunk 2]
```diff
@@ -74,22 +73,19 @@ public static  <K> void build(
...
-        schema,
-        timestampColumn
-    );
+    final Optional<TimestampProcessorSupplier<K>> tsProcessorSupplier
+        = createTimeStampProcessorSupplier(
+            buildContext, queryContext, schema, timestampColumn);
 
-    final KStream<K, GenericRow> transformed = tsTransformer
-        .map(t -> stream.transform(t, Named.as(TIMESTAMP_TRANSFORM_NAME
+    final KStream<K, GenericRow> transformed = tsProcessorSupplier
```
Comment (mjsax) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SinkBuilder.java:80]: Nit: should `transformed` be renamed?

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/process/KsProcessorSupplier.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  public KsProcessorSupplier(
+      final KsqlTransformer<KInT, KOutT> keyDelegate,
+      final KsqlTransformer<KInT, GenericRow> valueDelegate) {
+    this.processor = new KsProcessor<>(keyDelegate, valueDelegate);
+  }
+
+  @Override
+  public Processor<KInT, GenericRow, KOutT, GenericRow> get() {
+    return processor;
```
Comment (mjsax) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/process/KsProcessorSupplier.java:36]: This is not right. A supplier needs to return a new instance each time `get()` is called. You cannot use a singleton pattern and return the same object each time (if this would be allowed, we would not need to to use a supplier pattern to begin with...)

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/process/KsProcessor.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  @Override
+  public void process(final Record<KInT, GenericRow> record) {
+    final KInT key = record.key();
+    final GenericRow value = record.value();
+    final Record<KOutT, GenericRow> newRecord = new Record<>(
+        keyDelegate.transform(key, value, context),
+        valueDelegate.transform(key, value, context),
+        record.timestamp()
+    );
+    apiProcessorContext.forward(newRecord);
```
Comment (mjsax) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/process/KsProcessor.java:54]: Would it be cleaner/simpler to add a `forward` to the new `KsqlProcessingContext` so we don't need to maintain two context objects inside this class?

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/process/KsProcessor.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  @Override
+  public void close() { }
```
Comment (mjsax) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/process/KsProcessor.java:58]: Why do we overwrite `close()` given it's empty?

File: ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/SinkBuilderTest.java [hunk 6]
```diff
@@ -191,16 +193,18 @@ public void shouldTransformTimestampRow() {
     when(timestampExtractor.extract(any(), any())).thenReturn(timestampColumnValue);
 
     // When
-    final Transformer<Struct, GenericRow, KeyValue<Struct, GenericRow>> transformer =
+    final Processor<Struct, GenericRow, Struct, GenericRow> transformer =
```
Comment (mjsax) [inline ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/SinkBuilderTest.java:196]:
```suggestion
    final Processor<Struct, GenericRow, Struct, GenericRow> processor =
```

File: ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/SinkBuilderTest.java [hunk 7]
```diff
@@ -191,16 +193,18 @@ public void shouldTransformTimestampRow() {
...
         getTransformer(timestampExtractor, processingLogger);
```
Comment (mjsax) [inline ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/SinkBuilderTest.java:197]:
```suggestion
        getProcessor(timestampExtractor, processingLogger);
```

**Review by mjsax — COMMENTED** [review]
Made an initial pass.

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/process/KsStreamProcessingContext.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+  private final ProcessingContext context;
+
+  public KsStreamProcessingContext(final ProcessingContext processingContext) {
+    this.context = processingContext;
+  }
+
+  @Override
+  public long getRowTime() {
+    return context.currentStreamTimeMs();
```
Comment (mjsax) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/process/KsStreamProcessingContext.java:37]: This does not look right.

In the old API, we did pass `key` and `value` into eg `transform(...)` and getting a record's timestamp was done via `context.timestamp()`.

In the new API, we do pass `Record` with contain the timestamp of the record already, and thus it can be accessed by `process()` directly (w/o the detour of using the `context`). The old `context.timestamp()` was removed w/o a replacement, and `currentStreamTimeMs()` does not return a record's timestamp, but well, as the names suggest "stream time".

Not sure why this does not make any tests fail... Maybe because "stream time" is `max(r1.ts, ... rN.ts)` over all processed records r1,...,rN, and thus is the same as `record.timestamp()` as long as there is no out-of-order data? Or, maybe because this is actually unused? Looking into the code base, even parent interface `KsqlProcessingContext` seems not to be use anywhere (well, it's passed around, but `getRowTime()` is never call except test code...)

Bottom line is: I am not even sure if we still need `KsStreamProcessingContext` helper class as all, given the new API. I believe, it was just added to make semantics more clear -- `getRowTime()` is a better name than `context.timestamp()` for the case we need to access the record timestamp; but it seems, this was added pre-mature but never needed.
